### PR TITLE
Reapply "[mlir] NFC: fix dependence of (Tensor|Linalg|MemRef|Complex) dialects on LLVM Dialect and LLVM Core in CMake build (#104832)"

### PR DESCRIPTION
**Pull request by christopherbate**
Reapply the commit 43b508566799751aa180f1eaaafc5be693f2f1ae with additional fixes for building with
BUILD_SHARED_LIBS=ON.

### CODE REVIEW

**Comment by aartbik** [comment]
LGTM for sparse part

File: mlir/lib/Dialect/MemRef/Transforms/CMakeLists.txt [hunk 1]
```diff
@@ -35,11 +35,12 @@ add_mlir_dialect_library(MLIRMemRefTransforms
...
   MLIRLoopLikeInterface
   MLIRMemRefDialect
+  MLIRMemRefUtils
   MLIRNVGPUDialect
   MLIRPass
   MLIRTensorDialect
   MLIRTransforms
   MLIRValueBoundsOpInterface
-  MLIRVectorDialect
+  MLIRVectorDialect
```
Comment (matthias-springer) [inline mlir/lib/Dialect/MemRef/Transforms/CMakeLists.txt:44]: nit: space

File: mlir/lib/Conversion/VectorToLLVM/CMakeLists.txt [hunk 2]
```diff
@@ -13,10 +13,11 @@ add_mlir_conversion_library(MLIRVectorToLLVM
   Core
 
   LINK_LIBS PUBLIC
-  MLIRArithDialect
+  MLIRArithAttrToLLVMConversion
+  MLIRArithDialect  
   MLIRLLVMCommonConversion
   MLIRLLVMDialect
-  MLIRMemRefDialect
+  MLIRMemRefDialect
```
Comment (matthias-springer) [inline mlir/lib/Conversion/VectorToLLVM/CMakeLists.txt:20]: nit: space

File: mlir/lib/Conversion/SCFToControlFlow/SCFToControlFlow.cpp [hunk 3]
```diff
@@ -379,8 +378,8 @@ LogicalResult ForLowering::matchAndRewrite(ForOp forOp,
   // llvm.loop_annotation attribute.
   SmallVector<NamedAttribute> llvmAttrs;
   llvm::copy_if(forOp->getAttrs(), std::back_inserter(llvmAttrs),
-                [](auto attr) {
-                  return isa<LLVM::LLVMDialect>(attr.getValue().getDialect());
+                [](NamedAttribute attr) {
+                  return attr.getValue().getDialect().getNamespace() == "llvm";
```
Comment (matthias-springer) [inline mlir/lib/Conversion/SCFToControlFlow/SCFToControlFlow.cpp:382]: Why is this needed? Can `MLIRSCFToControlFlow` depend on the LLVM dialect?

Comment (joker-eph) [reply mlir/lib/Conversion/SCFToControlFlow/SCFToControlFlow.cpp:382]: Using a string here looks like a bit of a hack to me :(

It's pretty unfortunate that we have to choose between a dependency and the feature here though.

Comment (christopherbate) [reply mlir/lib/Conversion/SCFToControlFlow/SCFToControlFlow.cpp:382]: > Why is this needed? Can MLIRSCFToControlFlow depend on the LLVM dialect?

Well downstream users that don't use LLVM IR or LLVM MLIR Dialect may want to use this pass.   If downstream users are sensitive to the build/binary cost then they will have to duplicate this pass to eliminate the dependency. 

I guess it's not a huge deal to solve; we can expose the patterns as a separate function and have a callback to customize how attributes are propagated. The current behavior can remain the default and  downstream user can then instantiate a different pass without the LLVM dependency.

Comment (christopherbate) [reply mlir/lib/Conversion/SCFToControlFlow/SCFToControlFlow.cpp]: removed changes and will follow up with workaround mentioned above
